src/opencl/cryptmd5_kernel.cl

```diff
@@ -527,23 +631,39 @@ __kernel void cryptmd5(__global const crypt_md5_password *inbuffer,
 	ctx_buflen[7] += 16;
 	PUTCHAR(ctx[7].buffer, ctx_buflen[7], 0x80);
 
-#define altpos salt.c
-#ifdef UNROLL_LESS
-	salt.w[0] = 0;
+#ifndef UNROLL_LESS
+	uint ctx_buflen1, ctx_buflen2;
+	ctx_buflen1 = ctx_buflen[0] << 2;
+	ctx_buflen1 |= (uint)ctx_buflen[1] << 10;
+	ctx_buflen1 |= (uint)ctx_buflen[2] << 18;
+	ctx_buflen1 |= (uint)ctx_buflen[3] << 26;
+	ctx_buflen2 = ctx_buflen[4] << 2;
+	ctx_buflen2 |= (uint)ctx_buflen[5] << 10;
+	ctx_buflen2 |= (uint)ctx_buflen[6] << 18;
+	ctx_buflen2 |= (uint)ctx_buflen[7] << 26;
 #endif
-	altpos[4] = pass_len;
-	altpos[5] = pass_len * 2;
-	altpos[6] = pass_len + salt_len;
-	altpos[7] = altpos[5] + salt_len;
+
+	uint altpos;
+	altpos = pass_len;
+	altpos |= (pass_len * 2) << 8;
+	altpos |= (pass_len + salt_len) << 16;
+	altpos |= (pass_len * 2 + salt_len) << 24;
+
+/*
+ * In our uses of "altpos >> id2" below, we rely on OpenCL's defined behavior
+ * (unlike C99's undefined behavior) of right shifts by greater than the type's
+ * width.  In OpenCL, they shift by the number of bits encoded in the least
+ * significant bits of the shift count.
+ */
```

Comment:
This was news to me, but I'm yet to digest it 😄

Reply:
I actually looked this up in OpenCL 1.0 spec to make sure this dates back to there - it does.

Reply:
So if I shift a 16-bit integer with 19, does that mean it will shift by `19 & 15` so 3 - and not end up with a surefire zero? Anyway it's a good thing it's defined.

Reply:
Exactly. I think it's undefined in C for historical reasons since behavior of the corresponding machine instructions differs between archs (and is sometimes undefined as well, to allow it to differ between implementations of those archs, ouch). It's handy to have it defined in OpenCL, and shouldn't be a problem for GPUs, but I guess it does mean that OpenCL code using variable shifts may sometimes be slower than C code on CPUs.